### PR TITLE
Add linear regression 

### LINKED ISSUE #2: Explore linear regression models
We want to explore [linear **regression** models](https://scikit-learn.org/stable/modules/linear_model.html) on hybrid labs data, more specifically check [multi-output regressors](https://scikit-learn.org/stable/modules/multiclass.html#multioutput-regression). For that, a Python module needed to be called to fit, and predict. The module should:
- [ ] similar to [pycaret](https://pycaret.gitbook.io/docs) to compare different models, see [examples](https://nbviewer.org/github/pycaret/pycaret/blob/master/tutorials/Tutorial%20-%20Regression.ipynb)
- [ ] have a notebook in `docs` folder to show a simple example
- [ ] have tests
- [ ] have an easy interface and run fast (to be discussed)

**Data:**
The data folder `exp699_032024_TUDelft` is available on **Atlas** SharePoint at `Documents > HybridLabs > Example_data`. **NOTE:** The data folder `exp699_032024_TUDelft` is **not** public and cannot be shared with others. The trained model on this data cannot be also shared. For now, no need to store the trained model. 

The data folder `exp699_032024_TUDelft` includes:

- `readme.md`: it contains the experiment details and ML input/output
- `channel.csv:` it contains the variable names and units 
- `exp699.mat:` it contains the data and it is in matlab format.

For reading data in python, see [this notebook](https://github.com/hybridlabs-nl/FOWT-ML/blob/main/docs/notebooks/compare_linear_models.ipynb).

**Literature:**

Some literature is available at on **Atlas** SharePoint at `Documents > HybridLabs > Literature`. The two most related to FOWT are:

- [A hybrid methodology for wind tunnel testing of floating offshore wind turbines](https://doi.org/10.1016/j.oceaneng.2020.107592)
- [A review of modelling techniques for floating offshore wind turbines](https://doi.org/10.1002/we.2701)

### LINKED ISSUE #19: SonarCloud quality gate failing
The sonarcloud does not calculate the coverage correctly. Perhaps it does not find the path, see https://github.com/marketplace/actions/official-sonarqube-scan

### PR DESCRIPTION
**Pull request by SarahAlidoost**
closes #2 
closes #19

### CODE REVIEW

File: src/fowt_ml/linear_models.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+from dataclasses import dataclass
+import sklearn.linear_model as lm
+import sklearn.metrics as sm
+from numpy.typing import ArrayLike
+
+
+@dataclass
+class Estimator:
+    name: str
+    func: callable
```
Comment (fnattino) [inline src/fowt_ml/linear_models.py:12]: If I understand correctly, this is actually assigned an estimator instance (something that implements `estimator.fit()` and `estimator.predict()`) - so not a callable function, right?

File: src/fowt_ml/linear_models.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        "LeastAngleRegression": lm.Lars(),
+    }
+    METRICS_NAMES = {
+        "root_mean_squared_error": sm.root_mean_squared_error,
+        "mean_squared_error": sm.mean_squared_error,
+        "r2_score": sm.r2_score,
+        "mean_absolute_error": sm.mean_absolute_error,
+    }
+
+    def __init__(self, estimator: str | Estimator, kwargs: dict = None) -> None:
```
Comment (fnattino) [inline src/fowt_ml/linear_models.py:40]: Maybe instead of `kwargs` either use `**kwargs` (i.e. with unpacking) or use a more descriptive name (maybe `params`?)?

Also, I am not 100% sure, but if a parameter has `None` as default, this should be made explicit in the type hint:
```suggestion
    def __init__(self, estimator: str | Estimator, kwargs: dict | None = None) -> None:
```

File: src/fowt_ml/linear_models.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        elif isinstance(estimator, Estimator):
+            self.estimator = estimator
+            # TODO: validate if model function is a callable and valid
+        else:
+            raise ValueError("model must be a string or a Estimator instance.")
+
+        # fill the estimator with the kwargs
+        if kwargs:
+            for key, value in kwargs.items():
+                setattr(self.estimator.func, key, value)
```
Comment (fnattino) [inline src/fowt_ml/linear_models.py:60]: Estimator objects in scikit-learn should have a `set_params` method (see [link](https://scikit-learn.org/dev/developers/develop.html#estimators)), so maybe one could do:
```suggestion
        self.estimator.func.set_params(**kwargs)
```

**Review by fnattino — COMMENTED** [review]
Looks good @SarahAlidoost, just a few minor comment. 

One think that I have thought of right now while looking at the code is data normalization, which I think is not implemented but maybe should be included as part of the data preparation (maybe in the pipeline)?

File: src/fowt_ml/linear_models.py (hunk 1 above)
Comment (SarahAlidoost) [reply src/fowt_ml/linear_models.py]: I refactored this. Now it uses `BaseEstimator`

File: src/fowt_ml/linear_models.py (hunk 2 above)
Comment (SarahAlidoost) [reply src/fowt_ml/linear_models.py]: it is fixed now.

File: src/fowt_ml/linear_models.py (hunk 3 above)
Comment (SarahAlidoost) [reply src/fowt_ml/linear_models.py]: cool, I didn't know the method. thanks for the suggestion, it is fixed now.

**Comment by SarahAlidoost** [comment]
> Looks good @SarahAlidoost, just a few minor comment.
> 
> One think that I have thought of right now while looking at the code is data normalization, which I think is not implemented but maybe should be included as part of the data preparation (maybe in the pipeline)?

Thanks for the review and suggestions. I addressed your comments, mainly the implementation is refactored in line with random forest implementation. Please let me know if we can merge this PR.

File: src/fowt_ml/linear_models.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            x_train (ArrayLike): training data for features
+            x_test (ArrayLike): test data for features
+            y_train (ArrayLike): training data for targets
+            y_test (ArrayLike): test data for targets
+            scoring (Union[str, scoring]): the scoring to calculate
+
+        Returns:
+            float: the scoring value
+        """
+        self.model = self.estimator.fit(x_train, y_train)
```
Comment (fnattino) [inline src/fowt_ml/linear_models.py:63]: As I also commented in #16, I think `estimator.fit()` updates the estimator in place and then self-return - so I think `self.model` and `self.estimator` here would be identical..

**Review by fnattino — APPROVED** [review]
Looks good @SarahAlidoost  - I have left only one comment, feel free to address it or not  and merge it afterwards :)
